flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,738 @@
+        wd_min (float): The minimum wind direction value. Defaults to 0.0.
+        wd_max (float): The maximum wind direction value. Defaults to 360.0.
+        ws_step (float): The step size for the wind speed bins. Defaults to 1.0.
+        ws_min (float): The minimum wind speed value. Defaults to 0.0.
+        ws_max (float): The maximum wind speed value. Defaults to 50.0.
+        bin_cols_in (List[str]): A list of column names to bin the dataframes by.
+            Defaults to ["wd_bin", "ws_bin"].
+        weight_by (str): The method to weight the bins. Defaults to "min".
+        df_freq_pl (pl.DataFrame): A polars dataframe with the frequency of each bin.
+            Defaults to None.
+        percentiles (List[float]): The percentiles to for confidence intervals.
+            Defaults to [2.5, 97.5].
+        remove_any_null_turbine_bins (bool): When computing farm power, remove any bins where
+            and of the test turbines is null.  Defaults to False.
+        variance_only (bool): Only use the variance of turbine powers and not turbine-turbine
+            covariances. Defaults to False.
+        fill_cov_with_var (bool): Fill all missing covariance terms with the product of the
+            square root of the variances. Defaults to False.
+
+    Returns:
+        Dict[str, Dict[str, float]]: A dictionary containing the uplift results with standard error.
+    """
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+    #     print(df_sum)
+    #     print(uplift_results)
+
+    # Get the bin cols without df_name
+    bin_cols_without_df_name = [c for c in bin_cols_in if c != "df_name"]
+    bin_cols_with_df_name = bin_cols_in + ["df_name"]
+
+    # Add the wd and ws bins
+    df_ = _add_wd_ws_bins(
+        df_,
+        wd_cols=wd_cols,
+        ws_cols=ws_cols,
+        wd_step=wd_step,
+        wd_min=wd_min,
+        wd_max=wd_max,
+        ws_step=ws_step,
+        ws_min=ws_min,
+        ws_max=ws_max,
+    )
+
+    # Compute the covariance frame
+    df_cov = _compute_covariance(
+        df_, test_cols=test_cols, bin_cols_with_df_name=bin_cols_with_df_name
+    )
+
+    # If filling missing covariance terms, do it now
+    if fill_cov_with_var:
+        df_cov = _fill_cov_null(df_cov, test_cols=test_cols)
+
+    # If only using the variance, zero out the covariance terms
+    if variance_only:
+        df_cov = _zero_cov(df_cov, test_cols=test_cols)
+
+    # Apply Null values to covariance and sync across uplift pairs
+    df_cov = _null_and_sync_covariance(
+        df_cov=df_cov,
+        test_cols=test_cols,
+        uplift_pairs=uplift_pairs,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_cov)
+
+    # Bin and group the dataframe
+    df_bin = _bin_and_group_dataframe_expected_power(
+        df_=df_,
+        test_cols=test_cols,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
```

Comment:
I was thinking this could be a good place to call a new utility function that synchronizes nulls between df_cov and df_bin (something like "_synchronize_mean_power_cov_nulls"). Specifically, for each row, if there are any turbines in df_cov with undefined variances or covariances (because count < 2), then the mean power for those turbines would get set to Null as well. This way, we would always be able to return a standard error for the uplift by excluding turbines in a given bin with undefined covariance from both the expected uplift and uncertainty calculations. I like this because we wouldn't end up returning a NaN uncertainty value when calling this function. In most cases, though, we would probably just want to set `variance_only = True` or `fill_cov_with_var = True` to maximize the number of turbines that can be used to compute the uplift in each bin.